claws-mail: ensure latest plugins are used

oxzi opened the pull request:
## Description of changes
Our claws-mail package contains all built plugins next to the (already wrapped) claws-mail binary.

The user's `~/.claws-mail/clawsrc` configuration file is being generated through Claws Mail's GUI and might references the loaded plugins by their full path. Please note, older versions of Claws Mail seems to have
written only the plugin's basename to the config.

After a package update, the new claws-mail binary will be used, but the configuration might still points to the old plugins by their absolute path. This will result in either outdated and potentially vulnerable plugins being loaded or Claws Mail complaining about their absence after the Nix Store was cleaned by its Garbage Collection.

This patch uses the already implemented `plugin_load_in_default_dir` function from Claws Mail when loading
the plugins from the configuration file. By doing so, Claws Mail looks for a plugin with the basename from
the configuration in the current plugin directory.

Testing this with different configurations seems to work.

Loading the valid plugin `pgpcore.so` specified in any of the following variants succeeded with the same debug log below.
- Base name- `pgpcore.so`
- Correct absolute path: `/nix/store/rk7yfmmsac6lll1qdl6vlxyamhnl9cpy-claws-mail-4.2.0/lib/claws-mail/plugins/pgpcore.so`
- Outdated absolute path: `/nix/store/2wkg13m98n59sh5m9x8757x5c3ka49cn-claws-mail-4.2.0/lib/claws-mail/plugins/pgpcore.so`
- Invalid absolute path: `/nix/store/AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA-claws-mail-4.2.0/lib/claws-mail/plugins/pgpcore.so`

The debug output was always:
> plugin.c:390:trying to load pgpcore.so in default plugin directory /nix/store/rk7yfmmsac6lll1qdl6vlxyamhnl9cpy-claws-mail-4.2.0/lib/claws-mail/plugins/
> plugin.c:454:trying to load `/nix/store/rk7yfmmsac6lll1qdl6vlxyamhnl9cpy-claws-mail-4.2.0/lib/claws-mail/plugins/pgpcore.so'
> [...]
> plugin.c:528:Plugin PGP/Core (from file /nix/store/rk7yfmmsac6lll1qdl6vlxyamhnl9cpy-claws-mail-4.2.0/lib/claws-mail/plugins/pgpcore.so) loaded

However, specifying an absent plugin, e.g., `pgpgore.so`, will result in silently aborting the loading procedure, as Claws Mail would do without the patch.

Closes #271741.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

fpletz commented:
IMHO modifying the users' configuration is a quite ugly hack to fix this. I would prefer to patch claws-mail itself to be able to load modules relative to its install directory. What do you think?

oxzi commented:
> IMHO modifying the users' configuration is a quite ugly hack to fix this. I would prefer to patch claws-mail itself to be able to load modules relative to its install directory. What do you think?

Yes, this would definitely be the more pure approach and I have mixed feelings regarding this proposed change. However, such a Claws Mail patch would be bigger than this small awk script. I will try looking into "how much bigger" this would actually be and then decide if I'm going to try it. Please feel free to ping me if nothing has happened for over a month.

ajs124 commented:
Actually, in my config, it only says:
```
[Plugins_GTK3]
address_keeper.so
attachwarner.so
libravatar.so
managesieve.so
notification.so
pgpcore.so
pgpinline.so
pgpmime.so
```
without any absolute paths. I wonder why that is/how I did that.

fpletz commented:
Interesting, I also only have plugin filenames rather than absolute paths in my config like @ajs124 and it just works.

As far as I can see my config is from 2010 and created on another Linux distribution. I just tested loading a new plugin and it does indeed add it with an absolute path on NixOS. I wonder if I really didn't mess with plugins in claws-mail since I use NixOS though, since I started with NixOS in 2015 and have never gone back. :face_with_spiral_eyes:

oxzi committed:
claws-mail: ensure latest plugins are used

Our claws-mail package contains all built plugins next to the (already
wrapped) claws-mail binary.

The user's `~/.claws-mail/clawsrc` configuration file is being generated
through Claws Mail's GUI and might references the loaded plugins by
their full path. Please note, older versions of Claws Mail seems to have
written only the plugin's basename to the config.

After a package update, the new claws-mail binary will be used, but the
configuration might still points to the old plugins by their absolute
path. This will result in either outdated and potentially vulnerable
plugins being loaded or Claws Mail complaining about their absence after
the Nix Store was cleaned by its Garbage Collection.

This patch uses the already implemented `plugin_load_in_default_dir`
function from Claws Mail when loading the plugins from the
configuration file. By doing so, Claws Mail looks for a plugin with the
basename from the configuration in the current plugin directory.

Testing this with different configurations seems to work.

Loading the valid plugin `pgpcore.so` specified in any of the following
variants succeeded with the same debug log below.
- Base name- `pgpcore.so`
- Correct absolute path: `/nix/store/rk7yfmmsac6lll1qdl6vlxyamhnl9cpy-claws-mail-4.2.0/lib/claws-mail/plugins/pgpcore.so`
- Outdated absolute path: `/nix/store/2wkg13m98n59sh5m9x8757x5c3ka49cn-claws-mail-4.2.0/lib/claws-mail/plugins/pgpcore.so`
- Invalid absolute path: `/nix/store/AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA-claws-mail-4.2.0/lib/claws-mail/plugins/pgpcore.so`

The debug output was always:
> plugin.c:390:trying to load pgpcore.so in default plugin directory /nix/store/rk7yfmmsac6lll1qdl6vlxyamhnl9cpy-claws-mail-4.2.0/lib/claws-mail/plugins/
> plugin.c:454:trying to load `/nix/store/rk7yfmmsac6lll1qdl6vlxyamhnl9cpy-claws-mail-4.2.0/lib/claws-mail/plugins/pgpcore.so'
> [...]
> plugin.c:528:Plugin PGP/Core (from file /nix/store/rk7yfmmsac6lll1qdl6vlxyamhnl9cpy-claws-mail-4.2.0/lib/claws-mail/plugins/pgpcore.so) loaded

However, specifying an absent plugin, e.g., `pgpgore.so`, will result in
silently aborting the loading procedure, as Claws Mail would do without
the patch.

oxzi commented:
Thanks for your feedback, @ajs124 and @fpletz! Thanks to your configuration snippets with only the basename, I dared to look into Claws Mail's configuration parser and, as it turns out, it already has some logic for plugin loading outside the expected default path. Thus, it came down to a one-line-patch.

I have changed the commit message accordingly and updated the description above in the PR.

<details>
  <summary>Old description</summary>

Our claws-mail package contains all built plugins next to the (already wrapped) claws-mail binary.

The user's ~/.claws-mail/clawsrc configuration file is being generated through Claws Mail's GUI and references the loaded plugins by their full path. After a package update, the new claws-mail binary will be used,
but the configuration still points to the old plugins. This will result in either outdated and potentially
vulnerable plugins being loaded or Claws Mail complaining about their absence after the GC was used.

This fix adds another "--run" clause to the already used wrapper to substitute all loaded plugin paths on startup by a small awk script.
</details>

<details>
  <summary>New description</summary>

Our claws-mail package contains all built plugins next to the (already wrapped) claws-mail binary.

The user's `~/.claws-mail/clawsrc` configuration file is being generated through Claws Mail's GUI and might references the loaded plugins by their full path. Please note, older versions of Claws Mail seems to have
written only the plugin's basename to the config.

After a package update, the new claws-mail binary will be used, but the configuration might still points to the old plugins by their absolute path. This will result in either outdated and potentially vulnerable plugins being loaded or Claws Mail complaining about their absence after the Nix Store was cleaned by its Garbage Collection.

This patch uses the already implemented `plugin_load_in_default_dir` function from Claws Mail when loading
the plugins from the configuration file. By doing so, Claws Mail looks for a plugin with the basename from
the configuration in the current plugin directory.

Testing this with different configurations seems to work.

Loading the valid plugin `pgpcore.so` specified in any of the following variants succeeded with the same debug log below.
- Base name- `pgpcore.so`
- Correct absolute path: `/nix/store/rk7yfmmsac6lll1qdl6vlxyamhnl9cpy-claws-mail-4.2.0/lib/claws-mail/plugins/pgpcore.so`
- Outdated absolute path: `/nix/store/2wkg13m98n59sh5m9x8757x5c3ka49cn-claws-mail-4.2.0/lib/claws-mail/plugins/pgpcore.so`
- Invalid absolute path: `/nix/store/AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA-claws-mail-4.2.0/lib/claws-mail/plugins/pgpcore.so`

The debug output was always:
> plugin.c:390:trying to load pgpcore.so in default plugin directory /nix/store/rk7yfmmsac6lll1qdl6vlxyamhnl9cpy-claws-mail-4.2.0/lib/claws-mail/plugins/
> plugin.c:454:trying to load `/nix/store/rk7yfmmsac6lll1qdl6vlxyamhnl9cpy-claws-mail-4.2.0/lib/claws-mail/plugins/pgpcore.so'
> [...]
> plugin.c:528:Plugin PGP/Core (from file /nix/store/rk7yfmmsac6lll1qdl6vlxyamhnl9cpy-claws-mail-4.2.0/lib/claws-mail/plugins/pgpcore.so) loaded

However, specifying an absent plugin, e.g., `pgpgore.so`, will result in silently aborting the loading procedure, as Claws Mail would do without the patch.
</details>

Please feel free to take another look.
